drivers/xen/gnttab.c

```diff
@@ -0,0 +1,321 @@
+
+	rc = gnttab_reset_flags(gref);
+	if (!rc) {
+		return rc;
+	}
+
+	put_free_entry(gref);
+
+	return 0;
+}
+
+int32_t gnttab_alloc_and_grant(void **map, bool readonly)
+{
+	void *page;
+	unsigned long gfn;
+	grant_ref_t gref;
+
+	__ASSERT_NO_MSG(map != NULL);
+
+	page = k_aligned_alloc(XEN_PAGE_SIZE, XEN_PAGE_SIZE);
+	if (page == NULL) {
+		return -ENOMEM;
+	}
+
+	gfn = xen_virt_to_gfn(page);
+	gref = gnttab_grant_access(0, gfn, readonly);
+
+	*map = page;
+
+	return gref;
+}
+
+static void gop_eagain_retry(int cmd, struct gnttab_map_grant_ref *gref)
+{
+	unsigned int step = 10, delay = step;
+	int16_t *status = &gref->status;
+
+	do {
+		HYPERVISOR_grant_table_op(cmd, gref, 1);
+		if (*status == GNTST_eagain)
+			k_sleep(K_MSEC(delay));
+
+		delay += step;
+	} while ((*status == GNTST_eagain) && (delay < GOP_RETRY_DELAY));
+
+	if (delay >= GOP_RETRY_DELAY) {
+		LOG_ERR("Failed to map grant, timeout reached\n");
+		*status = GNTST_bad_page;
+	}
+}
+
+int gnttab_map_refs(struct gnttab_map_grant_ref *map_ops, unsigned int count)
+{
+	int i, ret;
+	struct xen_remove_from_physmap rfpm = {
+		.domid = DOMID_SELF,
+		.gpfn = xen_virt_to_gfn(map_ops->host_addr),
+	};
+
+	/*
+	 * GNTTABOP_map_grant_ref will simply replace the entry in the P2M
+	 * and not release any RAM that may have been associated with
+	 * map_ops->host_addr, so we release this memory before mapping.
+	 */
+	ret = HYPERVISOR_memory_op(XENMEM_remove_from_physmap, &rfpm);
+	if (ret) {
+		return ret;
+	}
+
+	ret = HYPERVISOR_grant_table_op(GNTTABOP_map_grant_ref, map_ops, count);
+	if (ret) {
+		return ret;
+	}
+
+	for (i = 0; i < count; i++) {
+		switch (map_ops[i].status) {
+		case GNTST_no_device_space:
+			LOG_WRN("map_grant_ref failed, no device space for page #%d\n", i);
+			break;
+
+		case GNTST_eagain:
+			/* Operation not done; need to try again */
+			gop_eagain_retry(GNTTABOP_map_grant_ref, &map_ops[i]);
+			/* Need to re-check status for current page */
+			i--;
+
+			break;
+
+		default:
+			break;
+		}
+	}
+
+	return 0;
+}
+
+int gnttab_unmap_refs(struct gnttab_map_grant_ref *unmap_ops, unsigned int count)
+{
+	int i, rc = 0;
+	struct xen_memory_reservation reservation;
+	xen_pfn_t extents[count];
+
+	rc = HYPERVISOR_grant_table_op(GNTTABOP_unmap_grant_ref, unmap_ops, count);
+	if (rc) {
+		return rc;
+	}
+
+	/*
+	 * After unmapping there will be a 4Kb holes in address space
+	 * at unmap_ops[i].host_addr positions. To keep it contiguous
+	 * and be able to return such addresses to memory allocator
+	 * we need to populate memory on unmapped positions here.
+	 */
+	for (i = 0; i < count; i++) {
+		extents[i] = xen_virt_to_gfn(unmap_ops[i].host_addr);
+	}
+	memset(&reservation, 0, sizeof(reservation));
+	reservation.domid = DOMID_SELF;
+	reservation.extent_order = 0;
+	reservation.nr_extents = count;
+	set_xen_guest_handle(reservation.extent_start, extents);
+
+	return HYPERVISOR_memory_op(XENMEM_populate_physmap, &reservation);
```

Comment:
As I wrote in February [1], I am still not convinced that we should call ``XENMEM_populate_physmap`` here.

Most of the PV protocol doesn't support persistent grant. So the backend will have to map/unmap the grant every time. In this case, we will also have to (un)populate the memory which is quite expensive.

I understand that mapping outside actual RAM (or using ballooning) would be more involved. However, I think we should at least get the interface right from the start. IOW, we should have two helpers:
  - Helper 1 will allocate memory and unpopulate the region
  - Helper 2 will populate the region and allocate the memory
  
 With this approach, it would be easy to change which pages are allocated without having to rework all the backends.

[1] https://github.com/zephyrproject-rtos/zephyr/pull/40969/commits/f6ef01138ba2cb25bc075fef5a697df4b859b429#r805319978

Reply:
@jgrall, I'm very sorry for late reply, I accidentally made comments with a "Start review" button and all of them were "Pending", please check them out

Reply:
> I probably got your point. Just to be sure, please tell me if I understood it right - you are okay with current approach, but you want to move all alloc/populate/unpopulate logic to separate functions (outside of gnttab_map/unmap_refs) to make it easier to implement other ways to obtain pages in future (e.g. ballooning)? If so I will fix it soon.

Correct.